### PR TITLE
Ec2 bugfixes

### PR DESCRIPTION
**Pull request by polvanrijn**
Includes following fixes:

- Check the instance name complies with same naming conventions as for the app name (underscores lead to errors)
- Allow to override `pem` and `security_group_name` (previously the passed parameters are always overridden by the .dallingerconfig)
- ~~Provide the instance name for adding (during provisioning) and removing (during teardown) the server name. E.g. if you do run this `dallinger ec2 provision --name abc --region eu-west-3 --dns-host xyz.my-experiments.com` you want to add the server `abc.xyz.my-experiments.com` and not just `xyz.my-experiments.com`, since your apps will be deployed to: `my_app.abc.xyz.my-experiments.com`~~ This was actually a mistake in my head, so nvm
- Fix a bug in fetching instance name if it's missing
- Bug fix for pem file stored as absolute path in config
- Add instance details like uptime, total cost, VCPU and memory

### CODE REVIEW

File: dallinger/command_line/ec2.py [hunk 1]
```diff
@@ -100,17 +100,15 @@ def list__instance_types(ctx, region):
...
 @click.option(
     "--image_name",
     default="ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-20230516",
     help="Image name; default is Ubuntu 22.04",
 )
 @click.option(
     "--security_group_name",
-    default="dallinger",
+    default=None,
     help="Security group name; default is dallinger",
```
Comment (pmcharrison) [inline dallinger/command_line/ec2.py:112]: Should mention that by default this comes from config.txt

File: dallinger/command_line/utils.py [hunk 2]
```diff
@@ -282,6 +282,14 @@ def verify_no_conflicts(verbose=True):
     return True
 
 
+def check_valid_subdomain(param_name, param_value):
+    """Check if the subdomain is valid."""
+    if not bool(re.match(r"^[a-z0-9-]+$", param_value)):
```
Comment (pmcharrison) [inline dallinger/command_line/utils.py:287]: Bool not necessary here

File: dallinger/command_line/lib/ec2.py [hunk 3]
```diff
@@ -684,7 +688,7 @@ def stop(region_name, instance_id):
     logger.info(f"Stopping of {instance_id} complete!")
 
 
-def teardown(region_name, instance_id, public_dns_name, dns_host):
+def teardown(region_name, instance_id, public_dns_name, dns_host, name=None):
```
Comment (pmcharrison) [inline dallinger/command_line/lib/ec2.py:691]: I would prefer the parameter to be called `instance_name` here as you have other names already in the call.

I am confused about why `name` is optional? Doesn't the user always have to specify it?

File: dallinger/command_line/ec2.py (hunk 1 above)
Comment (polvanrijn) [reply dallinger/command_line/ec2.py:112]: Updated the instructions now

File: dallinger/command_line/utils.py (hunk 2 above)
Comment (polvanrijn) [reply dallinger/command_line/utils.py:287]: Done

File: dallinger/command_line/lib/ec2.py (hunk 3 above)
Comment (polvanrijn) [reply dallinger/command_line/lib/ec2.py]: This part of the code has been modified

**Comment by polvanrijn** [comment]
@pmcharrison, I finished my work. You can have another pass if you like.

File: dallinger/command_line/ec2.py [hunk 4]
```diff
@@ -100,18 +100,16 @@ def list__instance_types(ctx, region):
...
     "--image_name",
     default="ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-20230516",
     help="Image name; default is Ubuntu 22.04",
 )
 @click.option(
     "--security_group_name",
-    default="dallinger",
-    help="Security group name; default is dallinger",
+    default=None,
+    help="Security group name; default is security_group_name from the config; fallback is dallinger",
```
Comment (pmcharrison) [inline dallinger/command_line/ec2.py:112]: Recommended wording:

```
Security group name; if not specified, defaults to the `security_group_name` config variable, whose default value is "dallinger".
```

File: dallinger/command_line/ec2.py [hunk 5]
```diff
@@ -100,18 +100,16 @@ def list__instance_types(ctx, region):
 @click.option("--region", default=None, help="Region name")
 @click.option("--type", default="m5.xlarge", help="Instance type")
 @click.option("--storage", default=32, type=int, help="Storage in GB; default is 32 GB")
-@click.option(
-    "--pem", default="dallinger", help="PEM file name; default is dallinger.pem"
-)
+@click.option("--pem", default=None, help="PEM file name; default is dallinger.pem")
```
Comment (pmcharrison) [inline dallinger/command_line/ec2.py:103]: Recommended wording for `help`:

Recommended wording:

```
Path to PEM file; if not specified, defaults to the `pem` config variable, whose default value is "dallinger.pem".
```

File: dallinger/command_line/lib/ec2.py [hunk 6]
```diff
@@ -79,21 +80,49 @@ def list_regions():
...
                     "instance_id": instance["InstanceId"],
                     "instance_type": instance["InstanceType"],
-                    "availability_zone": instance["Placement"]["AvailabilityZone"],
+                    "region": region_name,
                     "state": instance["State"]["Name"],
                     "public_dns_name": instance["PublicDnsName"],
                     # 'public_ip_address': instance['PublicIpAddress'],
                     "pem": instance["KeyName"],
+                    # Duration since instance was started
+                    "uptime": time.time() - instance["LaunchTime"].timestamp(),
```
Comment (pmcharrison) [inline dallinger/command_line/lib/ec2.py:125]: Doesn't this make an assumption about what the local timezone is?

File: dallinger/command_line/ec2.py (hunk 5 above)
Comment (fmhoeger) [inline dallinger/command_line/ec2.py:103]: Is the help message still correct if `default=None` now?

**Comment by pmcharrison** [comment]
Thanks @polvanrijn, maybe that's all the input I need to provide, but perhaps @fmhoeger can look through as well?

**Comment by pmcharrison** [comment]
It is quite low-risk though so don't let that hold you up.

**Comment by polvanrijn** [comment]
Cool, maybe you can approve it? So I can merge it later?

**Review by fmhoeger — COMMENTED** [review]
Just one comment

File: dallinger/command_line/ec2.py (hunk 4 above)
Comment (polvanrijn) [reply dallinger/command_line/ec2.py]: Addressed this.

File: dallinger/command_line/lib/ec2.py (hunk 6 above)
Comment (polvanrijn) [reply dallinger/command_line/lib/ec2.py]: Implemented your suggestion.

File: dallinger/command_line/ec2.py (hunk 5 above)
Comment (polvanrijn) [reply dallinger/command_line/ec2.py]: I updated it based on Peter's feedback

**Comment by fmhoeger** [comment]
@polvanrijn 
You might want to polish up the CHANGELOG description as I will copy it as is when doing the release.

**Comment by polvanrijn** [comment]
> @polvanrijn You might want to polish up the CHANGELOG description as I will copy it as is when doing the release.

Here are the lines for the changelog for commits:

**Fixed**
- Check the instance name complies with same naming conventions as for the app name
- Allow to override `pem` and `security_group_name`
- Fix a bug in fetching instance name if it's missing
- Bug fix for pem file stored as absolute path in config
- When starting a previously stopped EC2 instance, update the DNS entry
- Automatically restart all containers

**Added**
- Add instance details like uptime, total cost, VCPU and memory

**Review by fmhoeger — APPROVED** [review]

**Pull request merged**